### ISSUE TITLE
Batch-axis thread parallelism

### ISSUE THREAD

**Issue by staticfloat** [FluxML/Flux.jl#568]
On the `#helpdesk` slack channel, we got into a discussion about image augmentation, and how best to express that within Julia.  @adam-r-kowalski suggested an abstraction such as:

```
image = rand(RGB, 350, 250);

transform = Compose(
    Crop(300, 300),
    Pad(300, 300),
    ToChannels(Float32))

transform(image) # 300x300x3 Array{Float32, 3}
```

This feels very PyTorch-like, but then I thought to myself, why shouldn't these operations just be a part of the overall model?  There are really just two fundamental differences between the main model and "image preprocessing":

* You don't want `Zygote` to autodiff your jpeg loading code.  I think we can do this by just putting `Zygote.@nograd` in front of whatever we don't want tracked.

* You want multithreaded data parallelism across the batch axis of the data you're feeding into the network (I think you also want overlapped execution, but let's start simple for now).

I think it would be really useful for us to have a primitive like `BatchParallelMap(f, x)` that calls `f(x[...,thread_idx])` across a thread pool, then joins the results back together again (pending partr, I suppose).  With something like this, we would be able to have, for example, mulithreaded data loading where the data you pass into the Flux model is not tensors, but is instead filenames, for instance.  This isn't such a desirable thing in and of itself, I don't think, but it fits within the Flux ethos and worldview better than what we have now, IMO.

**Comment by MikeInnes** [comment]
Sure; these are just curried functions where `Crop(300, 300)(x) == crop(x, 300, 300)`. I think it's best to reuse functions from other packages where possible, but we can certainly add things like `BatchParallelMap` that don't exist elsewhere. Once we have the basic functionality we can also figure out whether the curried versions are more convenient to write (we added `MaxPool` after it got tiring to write `x -> maxpool(...)` in chains.

**Comment by KristofferC** [comment]
> There are really just two fundamental differences between the main model and "image preprocessing":

Isn't the third that you only want to do the preprocessing for the first epoch?

**Comment by datnamer** [comment]
There's this https://github.com/Evizero/Augmentor.jl

**Comment by staticfloat** [comment]
> Isn't the third that you only want to do the preprocessing for the first epoch?

The preprocessing often includes "augmentation" such as adding noise, random rotations, scaling, etc...., so in those cases it cannot be cached.

**Comment by ToucheSir** [comment]
This seems only tangentially related to Flux itself, though the topic of parallelism has come up quite a bit since. I think it's safe to close?

**Comment by CarloLucibello** [comment]
yeah, this should be discussed in https://github.com/lorenzoh/DataAugmentation.jl
